Use React hooks

gjvoosten opened the pull request:
Refactoring to use React hooks for asynchronous data fetching.

Partially addresses NCI-Agency/anet#1979

### User changes

- each section in the `Search` is responsible for querying and showing its own results (incl. pagination); each section will independently load (makes for a nice progressive experience), and is also responsible for showing 'No <objects> found'
- each `ReportCollection` visualisation (table, summary, calendar, map) is responsible for querying and showing the results (incl. pagination where relevant) meaning we now have `ReportTable`, `ReportSummary`, `ReportCalendar` and `ReportMap`, where the table and the summary retrieve paginated results, the calendar retrieves results only within the current view (e.g. month), and the map retrieves all results
- each `Home` tile is responsible for querying and showing its own results
- Excel exports have nicer worksheet names, and the mostly useless "first page of reports" worksheet is no longer present; export also limits the maximum number of results to 1000 per type/worksheet
- when printing the search results, the human-readable query stanza (like the one shown in the SearchBar) is now in the heading of the page (instead of just the query text)

maradragan commented on client/src/components/AdvisorReports/FilterableAdvisorReportsTable.js:
```diff
@@ -24,59 +23,52 @@ const GQL_GET_ADVISOR_REPORTS_INSIGHT = gql`
-    this.setState({ filterText: filterText })
+const FilterableAdvisorReportsTable = props => {
+  const [filterText, setFilterText] = useState("")
+  const [selectedData, setSelectedData] = useState([])
```
Somehow I was misleaded by the selectedData state, I thought it would be the filteredData. But it actually is the selectedRowsData. Right now if I use the filterText to filter the data and click export to CSV afterwards, without having checked any rows, I get all data in the export. I was expecting it to get only what I see. It is a bit strange, nothing is selected, I get everything back. But if I select rows I only get those ... It is not related to this pull request, maybe we should rethink if this is clear to the users.

gjvoosten replied:
Yes indeed, this is exactly how it behaves on candidate (I did not change the functionality). Please make a new issue to track this, @maradragan.

maradragan commented on client/src/components/AdvisorReports/AdvisorReportsTable.js:
```diff
@@ -1,13 +1,49 @@
+  }
+`
 
 const AdvisorReportsTable = props => {
```
Maybe we can agree to extract the props in the beginning of each function (like you did with the BarChart), so that we also know quicker which props it has? Now I can go to the end and look at the propTypes to see that.

gjvoosten replied:
In trying to minimise the changes (huh) I did not unfold the props everywhere. But as a convention we might adopt it. Could you make a debt issue to track this (and other JS style/conventions we want to follow)?

maradragan commented on client/src/components/AdvisorReports/FilterableAdvisorReportsTable.js:
```diff
@@ -24,59 +23,52 @@ const GQL_GET_ADVISOR_REPORTS_INSIGHT = gql`
   }
 
-  getWeekColumns() {
+  function getWeekColumns() {
```
Maybe it is nicer to have getWeekColumns outside the component.

gjvoosten replied:
Well, functions that are only used locally can just stay where they are (inside the function that uses them), trying to mimise the scope. It also makes it easier for eslint to spot functions that are unused (and thus can be removed).

maradragan commented on client/src/components/AdvisorReports/FilterableAdvisorReportsTable.js:
```diff
@@ -90,7 +82,7 @@ class FilterableAdvisorReportsTable extends Component {
   }
 
-  convertArrayOfObjectsToCSV(args) {
+  function convertArrayOfObjectsToCSV(args) {
```
Maybe it is nicer to have convertArrayOfObjectsToCSV outside the component.

gjvoosten replied:
Same here.

maradragan commented on client/src/components/AdvisorReports/FilterableAdvisorReportsTable.js:
```diff
@@ -140,9 +132,9 @@ class FilterableAdvisorReportsTable extends Component {
   }
 
-  downloadCSV(args) {
+  function downloadCSV(args) {
```
Maybe it is nicer to have downloadCSV outside the component.

gjvoosten replied:
Same here.

maradragan commented on client/src/components/BarChart.js:
```diff
@@ -204,44 +168,59 @@ class BarChart extends Component {
+
+  return <svg id={chartId} ref={node} width={width} height={height} />
+
+  function isNumeric(value) {
```
isNumeric could be placed outside the BarChart component.

gjvoosten replied:
Same here.

maradragan commented on client/src/components/Calendar.js:
```diff
@@ -1,78 +1,78 @@
-    )
-  }
+Calendar.defaultProps = {
+  events: []
```
This can be removed now right? Events are a required function now.

gjvoosten replied:
Yup, good point.

maradragan commented on client/src/pages/Home.js:
```diff
@@ -188,227 +272,185 @@ class BaseHome extends Page {
-          savedSearches: savedSearches,
-          selectedSearch: nextSelect
-        })
+        setSelectedSearch(undefined)
```
Why undefined and not null?

gjvoosten replied:
Remnant of old trials; I'll update it.

maradragan commented on client/src/pages/Search.js:
```diff
@@ -51,547 +54,834 @@ const GQL_CREATE_SAVED_SEARCH = gql`
+const Organizations = props => {
+  const {
+    queryParams,
+    setTotalCount,
```
I would give this property a different name, like handleSetTotalCount, or onSetTotalCount, just because now if I read setTotalCount , in combination with hooks, I would expect to have a useState hook for totalCount.

gjvoosten replied:
Well, actually, it *does* call a state hook setter, it's just that the setter is provided outside the component.

gjvoosten commented on client/src/components/CancelledEngagementReports.js:
```diff
@@ -1,395 +1,452 @@
+  selectedBarClass: PropTypes.string
+}
+
+const Collection = props => {
```
We could, but it would make the insight less self-contained.

gjvoosten commented on client/src/components/CancelledEngagementReports.js:
```diff
@@ -1,395 +1,452 @@
+  queryParams: PropTypes.object
+}
+
+const Map = props => {
```
We could, but it would make the insight less self-contained.

gjvoosten commented on client/src/components/DailyRollupChart.js:
```diff
@@ -212,30 +173,32 @@ class DailyRollupChart extends Component {
-    return true
+  return <svg id={chartId} ref={node} width={width} height={height} />
+
+  function isNumeric(value) {
```
It's a one-line function, although it is indeed quite generic.

gjvoosten commented on client/src/components/FutureEngagementsByLocation.js:
```diff
@@ -1,103 +1,313 @@
-      hoveredBar: null,
-      updateChart: true, // whether the chart needs to be updated
-      isLoading: false
+  function goToSelection(item) {
```
We could, but it would make the insight less self-contained.

gjvoosten commented on client/src/components/FutureEngagementsByLocation.js:
```diff
@@ -1,103 +1,313 @@
-      : ""
-    const focusedLocation = this.state.focusedSelection
-      ? this.state.focusedSelection.key
+  function updateHighlight(item, clear) {
```
We could, but it would make the insight less self-contained.

maradragan commented:
@gjvoosten If I create a new report, after clicking on save the blue loading indicator appears and stops somewhere near the end and stays on the page. This also happens when saving a new person, position ... didn't check further.
Also the loading indicator is quite slow, the save action already finished but the loading indicator still loads.

maradragan commented:
@gjvoosten On the admin page:
- I update a setting
- click on save
- see "admin settings saved"
- but the settings still has the old value
- if I go to the admin page again through the navigation I see the setting has the correct new value

gjvoosten commented:
> @gjvoosten On the admin page:
> 
>     * I update a setting
> 
>     * click on save
> 
>     * see "admin settings saved"
> 
>     * but the settings still has the old value
> 
>     * if I go to the admin page again through the navigation I see the setting has the correct new value

Fixed.

gjvoosten commented:
> @gjvoosten If I create a new report, after clicking on save the blue loading indicator appears and stops somewhere near the end and stays on the page. This also happens when saving a new person, position ... didn't check further.
> Also the loading indicator is quite slow, the save action already finished but the loading indicator still loads.

It is due to the history rewrite after saving from a create page: it causes the edit page to get loaded, and then we navigate to the show page. The loading of the edit page never gets the chance to finish (and doesn't need to; it's just so that Back from the Show page does not go to New but to Edit). Unfortunately, this also means the loading indicator oftentimes does not get cleared. I'm open to suggestions on solving this.

gjvoosten commented on client/src/components/ReportTable.js:
```diff
@@ -50,6 +175,53 @@ export default class ReportTable extends Component {
+    </>
+  )
+
+  function renderPagination() {
```
I've extracted all common pagination code into our `UltimatePagination` component.

gjvoosten commented:
> > @gjvoosten If I create a new report, after clicking on save the blue loading indicator appears and stops somewhere near the end and stays on the page. This also happens when saving a new person, position ... didn't check further.
> > Also the loading indicator is quite slow, the save action already finished but the loading indicator still loads.
> 
> It is due to the history rewrite after saving from a New page: it causes the Edit page to get loaded, and then we navigate to the Show page. The loading of the Edit page never gets the chance to finish (and doesn't need to; it's just so that Back from the Show page does not go to New but to Edit). Unfortunately, this also means the loading indicator oftentimes does not get cleared. I'm open to suggestions on solving this.

Well, commit https://github.com/NCI-Agency/anet/pull/2234/commits/1ec84bdfdcc07de5e7a16e978d9acd2d9748c06b seems to fix it.

gjvoosten commented on client/src/components/FutureEngagementsByLocation.js:
```diff
@@ -1,103 +1,313 @@
+      id: "febl-chart",
+      icons: [IconNames.HORIZONTAL_BAR_CHART],
+      title: "Chart by date and location",
+      renderer: getChart
```
Done.

maradragan commented:
When adding approval steps for an organization I get the following:
- if I add an incomplete approval step or an empty one, I can click n the save button of the organization but nothing happens.
- I would think I should get a validation error

maradragan commented:
On the Daily rollup page, I click on the end date, and while mousing over the calendar I see the start date becoming empty. When selecting a new end date I see the following error in the console:

TypeError: dateRange[0] is null
Show.js:479
changeRollupDate

maradragan commented:
@gjvoosten I go to the daily rollup page:
- I see "Rollup from startDate to EndDate" as title and underneath the 2 buttons for the types of orgs
- I click on an advisor org in the chart
- the title changes to "Rollup for xxx from startDate to endDate"
- the 2 buttons are replaced by one button: "All organizations"
- if I click on the  "All organizations" button the title changes to "Rollup for undefined from startDate to endDate"
- the 2 buttons for the types of orgs do not appear but the "All organizations" button stays on the page

maradragan commented:
@gjvoosten I went to the "Reports by day of the week" insight
- I go to the table view
- I click on the second page of the pagination and I see the expected reports
- I click in the chart on a day for which we have less than 10 reports
- I see in the table view No reports found

- The same happens with the summary view

maradragan commented:
@gjvoosten I go to the "Pending approval reports" insight
- when there are no reports pending approval the chart view displays one vertical line, the y-axis, and nothing more, I would expect it to at least also display an empty x-axis

maradragan commented:
@gjvoosten About the search :
- I search on EF9, I can find an organization with the name EF9
- I now search on EF, and I can also find EF9 on the 3rd page
- while being on the 3rd page of org results I search on EF9
- I see no results found, while I would expect it to find EF9

maradragan commented:
@gjvoosten Whe clicking on an item in the calendar the whole page gets empty, shows loading ... and than it's content gets filled ... this is a different experience for the user than when clicking on a read report link for instance from the summary view.

maradragan commented:
@gjvoosten 
- I go to an insights page
- I change the search filter and click on search
- I get redirected to the search results page, while I would expect it to filter the insight

This doesn't happen in candidate.

gjvoosten commented:
> @gjvoosten I go to the "Pending approval reports" insight
> 
>     * when there are no reports pending approval the chart view displays one vertical line, the y-axis, and nothing more, I would expect it to at least also display an empty x-axis

Reinstated showing _"No data"_ when the chart would be empty.

gjvoosten commented:
> @gjvoosten About the search :
> 
>     * I search on EF9, I can find an organization with the name EF9
> 
>     * I now search on EF, and I can also find EF9 on the 3rd page
> 
>     * while being on the 3rd page of org results I search on EF9
> 
>     * I see no results found, while I would expect it to find EF9
> 
> 
> This doesn't happen in candidate.

Should be fixed now; pagination is reset if search parameters change. Please check, @maradragan.

maradragan commented:
> @gjvoosten I go to the daily rollup page:
> 
>     * I see "Rollup from startDate to EndDate" as title and underneath the 2 buttons for the types of orgs
> 
>     * I click on an advisor org in the chart
> 
>     * the title changes to "Rollup for xxx from startDate to endDate"
> 
>     * the 2 buttons are replaced by one button: "All organizations"
> 
>     * if I click on the  "All organizations" button the title changes to "Rollup for undefined from startDate to endDate"
> 
>     * the 2 buttons for the types of orgs do not appear but the "All organizations" button stays on the page
> 
> 
> This doesn't happen in candidate.

This is fixed.

gjvoosten commented:
> @gjvoosten I went to the "Reports by day of the week" insight
> 
>     * I go to the table view
> 
>     * I click on the second page of the pagination and I see the expected reports
> 
>     * I click in the chart on a day for which we have less than 10 reports
> 
>     * I see in the table view No reports found
> 
>     * The same happens with the summary view
> 
> 
> This doesn't happen in candidate.

Should be fixed now, please check.

gjvoosten commented:
> @gjvoosten Whe clicking on an item in the calendar the whole page gets empty, shows loading ... and than it's content gets filled ... this is a different experience for the user than when clicking on a read report link for instance from the summary view.
> 
> This also happens in candidate, I will make an issue for it.

No need for a new issue, I fixed it in this branch/PR.

gjvoosten commented:
> @gjvoosten Whe clicking on an item in the calendar the whole page gets empty, shows loading ... and than it's content gets filled ... this is a different experience for the user than when clicking on a read report link for instance from the summary view.
> 
> This also happens in candidate, I will make an issue for it.

No need for a new issue, I fixed it in this branch/PR.

> @gjvoosten On the Daily rollup page, I click on the end date, and while mousing over the calendar I see the start date becoming empty. When selecting a new end date I see the following error in the console:
> 
> TypeError: dateRange[0] is null
> Show.js:479
> changeRollupDate
> 
> This also happens in candidate, I will make an issue for it.

Fixed in this branch/PR in commit 38f69766533f3536c04334f1b8bdf95443fe8763

gjvoosten commented on client/src/components/ReportCalendar.js:
```diff
@@ -0,0 +1,108 @@
+`
+
+const ReportCalendar = props => {
+  const { queryParams, setTotalCount } = props
```
There's nothing special about a state call, it's just a callback function, like this one, so I don't see a pressing need to change the name.

gjvoosten commented:
> @gjvoosten
> 
>     * I go to an insights page
> 
>     * I change the search filter and click on search
> 
>     * I get redirected to the search results page, while I would expect it to filter the insight
> 
> 
> This doesn't happen in candidate.

Fixed.
